src/context/component.js

```diff
@@ -0,0 +1,261 @@
+      deps.forEach((dep, index) =>
+        values.subscribe(dep, this.depSubscribers_[index])
+      );
+    }
+
+    // Run the first time.
+    if (this.isConnected_()) {
+      this.update_();
+    }
+  }
+
+  /**
+   * Called when the component is completely discarded, for instance via
+   * `removeComponent` API.
+   */
+  dispose() {
+    // Unsubscribe from all dependencies.
+    if (this.deps_.length > 0) {
+      const {values} = this.contextNode;
+      this.deps_.forEach((dep, index) =>
+        values.unsubscribe(dep, this.depSubscribers_[index])
+      );
+    }
+
+    this.cleanup_();
+  }
+
+  /**
+   * Called when the node's root has changed. It can be a different root or
+   * the node can be disconnected entirely.
+   */
+  rootUpdated() {
+    const isConnected = this.isConnected_();
+    this.cleanup_();
+    if (isConnected) {
+      this.update_();
+    }
+  }
+
+  /**
+   * Sets the component's input.
+   * @param {*} input
+   */
+  set(input) {
+    if (this.input_ !== input) {
+      this.input_ = input;
+      if (this.isConnected_()) {
+        this.update_();
+      }
+    }
+  }
+
+  /**
+   * Allocates a new reference in the component's internal state.
+   *
+   * See `useRef` hook for more info.
+   *
+   * @param {T} def
+   * @return {{current: T}}
+   * @template T
+   */
+  allocRef(def = undefined) {
+    const refs = this.refs_ || (this.refs_ = []);
+    const pointer = ++this.refPointer_;
+    return refs[pointer] || (refs[pointer] = {current: def});
+  }
+
+  /**
+   * Register a cleanup handler that will be called when the component is
+   * cleaned up, either due to the node being disconnected, or the component
+   * being removed.
+   *
+   * @param {function()} cleanup
+   */
+  pushCleanup(cleanup) {
+    const cleanups = this.cleanups_ || (this.cleanups_ = []);
+    pushIfNotExist(cleanups, cleanup);
+  }
+
+  /**
+   * Unregisters a cleanup handler previously registered with `pushCleanup`.
+   *
+   * @param {function()} cleanup
+   */
+  popCleanup(cleanup) {
+    const cleanups = this.cleanups_;
+    if (cleanups) {
+      removeItem(cleanups, cleanup);
+    }
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  isConnected_() {
+    return !!this.contextNode.root;
+  }
+
+  /** @private */
+  update_() {
+    if (!this.isConnected_()) {
+      // The node was disconnected at some point.
+      return;
+    }
+    const running = this.depValues_.every(isDefined);
+    if (running) {
+      this.running_ = true;
+      this.refPointer_ = -1;
+      withComponent(this, this.run_);
+    } else if (this.running_) {
+      this.running_ = false;
+      this.cleanup_();
+    }
+  }
+
+  /** @private */
+  run_() {
+    // Cleanup the previous run.
+    if (this.runCleanup_) {
+      protectedNoInline(this.runCleanup_);
+      this.runCleanup_ = null;
+    }
+
+    // Run the component.
+    const func = this.func_;
+    this.runCleanup_ = func(this, this.input_, this.depValues_);
```

Comment:
Why do `Component`s have a cleanup outside of hooks? Is it functionally equivalent to wrapping the entire contents of a component in a dependency-less `useEffect()`? (e.g. it runs each time and has a cleanup)

Reply:
Yeah. This is actually a code I added based on your and my thread on the design doc :). What we have is a component, but it's also somewhat closer to an effect, b/c it doesn't do any rendering. So I decided to make it a bit more like effect. In a way, the component structure now looks like this:

```
mainEffect {
  childEffect() {}
  childEffect() {}
  childEffect() {}
}
```

I didn't go as far as actually calling it an "effect" b/c I though it'd be more confusing to folk coming from React's experience where effects are flat. But it's somewhat of a cross between the two.

Reply:
The top level cleanup is basically for subscriber model to avoid extra allocations. E.g. 

```
subscribe(node, [Dep1, Dep2], (dep1, dep2) => {
  if (dep1 && dep2) {
    // Do some work.
    return function cleanup() {};
  }
})
```

This maps 1:1 to a simple component that simply calls the subscriber handler. Since I expected this to be a very frequent case, I wanted to support the top-level cleanup to make this 0-allocation case. But otherwise, I think, it should be natural for folk to keep the cleanup functions narrower to avoid excessive cleanups.

Reply:
> I don't think we'd want anything expensive that needs cleanups happening at top-level without bail-out conditions (like hooks have with the deps array).

Ah, right. The `subscribe` model _does_ have deps array. Makes sense. Thank you for all of the explanations!